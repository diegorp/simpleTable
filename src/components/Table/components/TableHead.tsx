import type { Column, Sort } from '@/types';
import { ARIA_SORT_VALUES } from '@/constants';

interface Props<T> {
  columns: Column<T>[];
  sort?: boolean;
  currentSort: Sort<T>;
  onToggleSort: (key: keyof T) => void;
}

export default function TableHead<T>({
  columns,
  sort,
  currentSort,
  onToggleSort,
}: Props<T>) {
  return (
    <thead>
      <tr>
        {columns.map(({ key, label }) => (
          <th
            key={key as string}
            scope='col'
            className='p-0 text-left text-xs font-medium text-gray-500 capitalize tracking-wider sticky top-0 z-10 shadow-[inset_0_-1px_0] shadow-[#E2E8F0] border-r last:border-r-0 border-[#E2E8F0] bg-white hover:bg-gray-100 transition-colors duration-200 ease-in-out'
            aria-sort={ARIA_SORT_VALUES[currentSort.direction]}
          >
            {sort ? (
              <button
                onClick={() => onToggleSort(key)}
                className='w-full h-full px-6 py-3 flex items-center gap-2 capitalize cursor-pointer text-left'
                aria-label={`Sort by ${label}`}
              >
                {currentSort.key === key && (
                  <span className='text-gray-500 text-xs'>
                    {currentSort.direction === 'asc'
                      ? '↑'
                      : currentSort.direction === 'desc'
                      ? '↓'
                      : ''}
                  </span>
                )}
                {label}
              </button>
            ) : (
              label
            )}
          </th>
        ))}
      </tr>
    </thead>
  );
}
