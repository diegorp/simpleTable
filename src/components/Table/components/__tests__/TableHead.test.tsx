import { render, fireEvent } from '@testing-library/react';
import TableHead from '../TableHead';
import type { Column, Sort } from '@/types';

describe('<TableHead />', () => {
  const mockColumns: Column<{ id: number; name: string; age: number }>[] = [
    { key: 'id', label: 'ID', render: (value) => String(value) },
    { key: 'name', label: 'Name', render: (value) => String(value) },
    { key: 'age', label: 'Age', render: (value) => String(value) },
  ];

  const mockSort: Sort<{ id: number; name: string; age: number }> = {
    key: 'id',
    direction: 'none',
  };

  it('should match snapshot', () => {
    const { container } = render(
      <table>
        <TableHead
          columns={mockColumns}
          sort={false}
          currentSort={mockSort}
          onToggleSort={vi.fn()}
        />
      </table>
    );
    expect(container).toMatchSnapshot();
  });

  it('should render all column headers', () => {
    const { getByText } = render(
      <table>
        <TableHead
          columns={mockColumns}
          sort={false}
          currentSort={mockSort}
          onToggleSort={vi.fn()}
        />
      </table>
    );

    expect(getByText('ID')).toBeInTheDocument();
    expect(getByText('Name')).toBeInTheDocument();
    expect(getByText('Age')).toBeInTheDocument();
  });

  it('should render sortable headers when sort is enabled', () => {
    const { getAllByRole } = render(
      <table>
        <TableHead
          columns={mockColumns}
          sort={true}
          currentSort={mockSort}
          onToggleSort={vi.fn()}
        />
      </table>
    );

    const buttons = getAllByRole('button');
    expect(buttons).toHaveLength(3);
  });

  it('should not render sortable headers when sort is disabled', () => {
    const { queryAllByRole } = render(
      <table>
        <TableHead
          columns={mockColumns}
          sort={false}
          currentSort={mockSort}
          onToggleSort={vi.fn()}
        />
      </table>
    );

    const buttons = queryAllByRole('button');
    expect(buttons).toHaveLength(0);
  });

  it('should call onToggleSort when header is clicked', () => {
    const onToggleSort = vi.fn();
    const { getByRole } = render(
      <table>
        <TableHead
          columns={mockColumns}
          sort={true}
          currentSort={mockSort}
          onToggleSort={onToggleSort}
        />
      </table>
    );

    const nameButton = getByRole('button', { name: 'Sort by Name' });
    fireEvent.click(nameButton);

    expect(onToggleSort).toHaveBeenCalledWith('name');
  });

  it('should display ascending sort indicator', () => {
    const ascSort: Sort<{ id: number; name: string; age: number }> = {
      key: 'name',
      direction: 'asc',
    };

    const { getByText } = render(
      <table>
        <TableHead
          columns={mockColumns}
          sort={true}
          currentSort={ascSort}
          onToggleSort={vi.fn()}
        />
      </table>
    );

    expect(getByText('↑')).toBeInTheDocument();
  });

  it('should display descending sort indicator', () => {
    const descSort: Sort<{ id: number; name: string; age: number }> = {
      key: 'name',
      direction: 'desc',
    };

    const { getByText } = render(
      <table>
        <TableHead
          columns={mockColumns}
          sort={true}
          currentSort={descSort}
          onToggleSort={vi.fn()}
        />
      </table>
    );

    expect(getByText('↓')).toBeInTheDocument();
  });

  it('should not display sort indicator for none direction', () => {
    const noneSort: Sort<{ id: number; name: string; age: number }> = {
      key: 'name',
      direction: 'none',
    };

    const { queryByText } = render(
      <table>
        <TableHead
          columns={mockColumns}
          sort={true}
          currentSort={noneSort}
          onToggleSort={vi.fn()}
        />
      </table>
    );

    expect(queryByText('↑')).not.toBeInTheDocument();
    expect(queryByText('↓')).not.toBeInTheDocument();
  });

  it('should only show sort indicator on sorted column', () => {
    const sortedByName: Sort<{ id: number; name: string; age: number }> = {
      key: 'name',
      direction: 'asc',
    };

    const { getAllByText } = render(
      <table>
        <TableHead
          columns={mockColumns}
          sort={true}
          currentSort={sortedByName}
          onToggleSort={vi.fn()}
        />
      </table>
    );

    const indicators = getAllByText('↑');
    expect(indicators).toHaveLength(1);
  });

  it('should handle empty columns array', () => {
    const { queryAllByRole } = render(
      <table>
        <TableHead
          columns={[]}
          sort={false}
          currentSort={mockSort}
          onToggleSort={vi.fn()}
        />
      </table>
    );

    const headers = queryAllByRole('columnheader');
    expect(headers).toHaveLength(0);
  });
});
