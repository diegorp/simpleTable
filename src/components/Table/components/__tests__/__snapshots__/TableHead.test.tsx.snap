// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<TableHead /> > should match snapshot 1`] = `
<div>
  <table>
    <thead>
      <tr>
        <th
          aria-sort="other"
          class="p-0 text-left text-xs font-medium text-gray-500 capitalize tracking-wider sticky top-0 z-10 shadow-[inset_0_-1px_0] shadow-[#E2E8F0] border-r last:border-r-0 border-[#E2E8F0] bg-white hover:bg-gray-100 transition-colors duration-200 ease-in-out"
          scope="col"
        >
          ID
        </th>
        <th
          aria-sort="other"
          class="p-0 text-left text-xs font-medium text-gray-500 capitalize tracking-wider sticky top-0 z-10 shadow-[inset_0_-1px_0] shadow-[#E2E8F0] border-r last:border-r-0 border-[#E2E8F0] bg-white hover:bg-gray-100 transition-colors duration-200 ease-in-out"
          scope="col"
        >
          Name
        </th>
        <th
          aria-sort="other"
          class="p-0 text-left text-xs font-medium text-gray-500 capitalize tracking-wider sticky top-0 z-10 shadow-[inset_0_-1px_0] shadow-[#E2E8F0] border-r last:border-r-0 border-[#E2E8F0] bg-white hover:bg-gray-100 transition-colors duration-200 ease-in-out"
          scope="col"
        >
          Age
        </th>
      </tr>
    </thead>
  </table>
</div>
`;
